os/drivers/video/video_null_driver.c

```diff
@@ -0,0 +1,1554 @@
+ *
+ * Input Parameters:
+ *   video_private - Reference to private data.
+ *   ctrl_class - V4L2 control class id
+ *   control - Pointer to control data to be set.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int video_null_set_ctrlvalue(FAR void *video_private, uint16_t ctrl_class, FAR struct v4l2_ext_control *control)
+{
+	int i = 0;
+	int32_t min;
+	int32_t max;
+	uint32_t step;
+	int ret = -EINVAL;
+	struct video_lowerhalf_s *phalf = (struct video_lowerhalf_s *)video_private;
+	FAR null_priv_t *priv = NULL;
+
+	if (phalf == NULL) {
+		return -EINVAL;
+	}
+
+	priv = (FAR null_priv_t *) phalf->dev;
+
+	if (priv == NULL) {
+		return -EINVAL;
+	}
+
+	video_null_takesem(&priv->sem);
+	for (i = 0; i < priv->ctrlcount; i++) {
+		if (priv->ctrl[i].ctrl_class == ctrl_class && priv->ctrl[i].id == control->id) {
+			if ((priv->ctrl[i].type == V4L2_CTRL_TYPE_INTEGER) || (priv->ctrl[i].type == V4L2_CTRL_TYPE_BOOLEAN)) {
+				/* Return if control value is not in range */
+				min = priv->ctrl[i].min;
+				max = priv->ctrl[i].max;
+				step = priv->ctrl[i].step;
+				ret = OK;
+			}
+			//TODO:: other types
+		}
+	}
+
+	if (ret == OK) {
+		video_null_givesem(&priv->sem);
+		CHECK_RANGE(control->value, min, max, step);
+		video_null_takesem(&priv->sem);
+		priv->ctrl[i].current_value = control->value;
+	}
+	video_null_givesem(&priv->sem);
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: video_null_refresh
+ *
+ * Description:
+ *   This function is video dev ops for refreshing the video stream.
+ *
+ *   This function is called from the video driver using dev ops.
+ *
+ * Input Parameters:
+ *   video_private - Reference to private data.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int video_null_refresh(FAR void *video_private)
+{
+	int ret = OK;
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: video_null_initialize
+ *
+ * Description:
+ *   This completes the initialization operations and register video driver
+ *
+ * Input Parameters:
+ *   devpath -  path to video device
+ *
+ * Returned Values:
+ *   None
+ *
+ ****************************************************************************/
+int video_null_initialize(const char *devpath)
+{
+	int ret = OK;
+
+	struct video_lowerhalf_s *lower = NULL;
+	struct dummy_null_priv_s *priv = NULL;
+
+	/* Allocate the null driver structure */
+	lower = (FAR struct video_lowerhalf_s *)
+			kmm_zalloc(sizeof(struct video_lowerhalf_s));
```

Comment:
Do you need zalloc? This structure has 2 elements and they are assigned at line 1491 and 1531.